### ISSUE TITLE
Type alias for model_dump include/exclude is broken

### ISSUE THREAD

**Issue by tommasolevato** [pydantic/pydantic#10333]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Version 2.9.0 changed the `IncEx` type alias to the following:

`IncEx: TypeAlias = Union[Set[int], Set[str], Dict[int, 'IncEx'], Dict[str, 'IncEx'], None]`

This is not correct for the `include` and `exclude` parameters of the `model_dump` method. For example, it's now not possible to use a `Dict[str, bool]` as the value for include/exclude (see the code example below).

### Example Code

```Python
from pydantic import BaseModel

class ExampleModel(BaseModel):
    example_field: str
    field_to_be_excluded: str

e = ExampleModel(example_field="Hello Pydantic", field_to_be_excluded="To be excluded")
e.model_dump(exclude={"field_to_be_excluded": True}) # mypy complains about this
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.9.0

pydantic-core version: 2.23.2

pydantic-core build: profile=release pgo=false

install path: /Users/*****/.venv/lib/python3.12/site-packages/pydantic

python version: 3.12.2 (main, Mar  7 2024, 08:27:42) [Clang 15.0.0 (clang-1500.3.9.4)]

platform: macOS-14.6.1-arm64-arm-64bit

related packages: mypy-1.11.2 fastapi-0.113.0 pydantic-extra-types-2.9.0 typing_extensions-4.12.2 pydantic-settings-2.4.0

commit: unknown
```

**Comment by sydney-runkle** [comment]
@tommasolevato,

Thanks for reporting. Indeed, looks like an issue in v2.9.0, we'll work to roll out a fix shortly.

**Comment by tommasolevato** [comment]
I've found another problem with `IncEx`, that v2.9.1 did not address:

```
class ExampleElement(BaseModel):
    number: int
    name: str


class ExampleModel(BaseModel):
    elements: list[ExampleElement]


e = ExampleModel(elements=[ExampleElement(number=1, name="Hello Pydantic")])
e.model_dump(exclude={"elements": {"_all": "number"}})  # mypy complains about this
```

I wasn't sure if it was appropriate to open another issue, so I decided to add a comment here. I can create a new issue if needed.

**Comment by Viicos** [comment]
Did you mean to write:

```python
e.model_dump(exclude={"elements": {"__all__": {"number"}}})
```

?

Afaik `dict[str, str]` isn't supported.

**Comment by tommasolevato** [comment]
Indeed, sorry for the confusion.

**Comment by sydney-runkle** [comment]
Hmm, did this work in previous pydantic versions? I don't recall us supporting something like this...

**Comment by tommasolevato** [comment]
Yes, ref: https://docs.pydantic.dev/latest/concepts/serialization/#advanced-include-and-exclude

**Comment by sydney-runkle** [comment]
@tommasolevato, gotcha - please open a new issue!

**Comment by Viicos** [comment]
@sydney-runkle I think @tommasolevato was talking about the `bool` type not being supported (which we fixed), so no need to open any new issue.

**Comment by tommasolevato** [comment]
@Viicos the `e.model_dump(exclude={"elements": {"__all__": {"number"}}})` example code is still broken in v2.9.1.

**Comment by Viicos** [comment]
It works on my end:

```py
e.model_dump(exclude={"elements": {"__all__": {"number"}}})
#> {'elements': [{'name': 'Hello Pydantic'}]}
```

**Comment by tommasolevato** [comment]
@Viicos operationally it works, but mypy complains about this, and I believe it's right in doing so.

**Comment by Viicos** [comment]
> mypy complains about this, and I believe it's right in doing so.

Ah yes, seems like there's an issue with variance. Oddly, pyright does not flag the error. I'll come back to you once I get an answer from pyright if it happens to be a false negative from pyright (and fix the type alias accordingly).

**Comment by bluenote10** [comment]
Just to clarify, this issue isn't resolved, right?

At least, a minimal example still doesn't type check with mypy in 2.9.1 / after #10339.

```py
from pydantic import BaseModel

class MyModel(BaseModel):
    foo: int
    bar: int

m = MyModel(foo=1, bar=2)
m.model_dump(include={"foo": True})
```
Error:
```
error: Argument "include" to "model_dump" of "BaseModel" has incompatible type "dict[str, bool]"; expected "IncEx | None"  [arg-type]
```

--------------

> Ah yes, seems like there's an issue with variance.

I also think so. I assume this line

https://github.com/pydantic/pydantic/blob/a6dc87285f93f90c2d5c298ee7c52f5d7e878194/pydantic/main.py#L73

should probably become
```py
IncEx: TypeAlias = Union[Set[int], Set[str], Mapping[int, Union['IncEx', bool]], Mapping[str, Union['IncEx', bool]]]
```
to make it covariant in the value type.

**Comment by Viicos** [comment]
Yes, we still have this issue closed as this is going to be tackled as part of https://github.com/pydantic/pydantic/issues/10335. The variance issue seems to be a mypy limitation/false positive, as per https://github.com/microsoft/pyright/discussions/8972#discussioncomment-10628016.

**Comment by bluenote10** [comment]
> The variance issue seems to be a mypy limitation/false positive, as per https://github.com/microsoft/pyright/discussions/8972#discussioncomment-10628016.

Note that Eric was referring to the local bidirectional inference. The type alias should generally type check against both `dict[str, bool]` and `dict[str, IncEx]` individually (and their immutable forms) if passed in externally, so using `Mapping` goes beyond that inference problem. For instance, this should type check, and has nothing to do with mypy's local inference limitation:

```py
# This function properly communicates that it will not mutate `include`.
def some_funct(x: MyModel, include: Mapping[str, bool]):
    print(x.model_dump_json(include=include))
```
